[CORE-671] Make bucket read-only in locked workspaces

samanehsan opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/CORE-671

When a workspace is locked:
- Update the workspace bucket permissions to read-only for all policy groups
- Remove any fast pass grants that exist. They directly grant the user and pet service accounts write access to the workspace project/bucket IAM because it can take some time for google group permissions to propagate. Note the grants are only valid for 2 hours, but if a user locks their workspace during that time and permissions have not propagated yet they will loose access until the policy group permissions finish propagating.

When a workspace is unlocked:
- Restore bucket permissions for all policy groups
- Re-add fast pass grants for this workspace so that the user can get write-access immediately

In the Terra UI and google console, users will not be able to upload data into the workspace bucket:
<img width="1373" height="392" alt="Screenshot 2025-09-18 at 1 17 16 PM" src="https://github.com/user-attachments/assets/11863de4-b9a1-4b74-9592-27d256010e51" />

---

**PR checklist**

- [x] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should publish a new official `rawls-model` and perform the corresponding dependency updates as specified in the [README](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model):
  - [ ] in the automation subdirectory
  - [ ] in workbench-libs
  - [ ] in firecloud-orchestration
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

samanehsan committed:
Make bucket read-only when locking a workspace

samanehsan committed:
Restore bucket IAM when unlocking workspace

samanehsan committed:
Fix compilation and test errors

samanehsan committed:
Add unit tests

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-671-update-lock-logic

samanehsan committed:
Scalafmt

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1603,10 +1603,34 @@ class WorkspaceService(
+      .flatten
+      .toMap
+
+    _ <- gcsDAO.updateBucketIam(GcsBucketName(workspace.bucketName), policyEmails, Option(workspace.googleProjectId))
```
`updateBucketIamAllReaders` put the Rawls service account into the custom reader role. Does this need to do anything to restore the service account's access? Does `updateBucketIamAllReaders` _need_ to put the service account in that role?

samanehsan replied:
Oh you're right I can leave the service account as-is

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -198,6 +198,36 @@ class HttpGoogleServicesDAO(val clientSecrets: GoogleClientSecrets,
+      .drain
+      .unsafeToFuture()
+  }
+
```
is this method necessary? When locking a workspace, could `WorkspaceService` build the role map and call `updateBucketIam` just like when unlocking a workspace?

I am fine either way; this question is about preference/style, just want to make sure you've considered it.

samanehsan replied:
The `policyGroupsByAccessLevel`map in `updateBucketIam` maps one role to one policy email but in this case we want both the owner and writer policy emails to have the reader role. I considered refactoring it to a `Map[WorkspaceAccessLevel, Set[WorkbenchEmail]]` but I thought the separate method made the intent clearer.

samanehsan committed:
Do not change bucket permissions for Rawls SA

davidangb reviewed (approved):
:+1:, assuming tests pass

calypsomatic commented:
I tested this out locally (just by calling the lock API from this branch) and it gave me this error: `Group policy-58fe6139-637e-4dd6-a5ae-b8470614a584@dev.test.firecloud.org does not exist`, otherwise seemed to have successfully locked the workspace, but I was still able to upload to the bucket.

samanehsan committed:
Fix overriding project owner bucket permissions

A different policy email is used depending on
whether the workspace has an auth domain.

samanehsan committed:
Fix iamPolicyVersion

If a policy uses v3 and Rawls tries to
override it with a v1 policy google will
throw an error. However if a v3 policy is
set but only a v1 policy is needed Google
will use that one.

samanehsan committed:
Handle fast pass permissions

When a workspace is created the user
and pet service account emails are
directly given write access to the bucket.
This change removes the grants from a locked
workspace and adds them back on unlock.

samanehsan commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -142,7 +142,7 @@ class HttpGoogleServicesDAO(val clientSecrets: GoogleClientSecrets,
                                policyGroupsByAccessLevel: Map[WorkspaceAccessLevel, WorkbenchEmail],
                                userProject: Option[GoogleProjectId],
-                               iamPolicyVersion: Int = 1
+                               iamPolicyVersion: Int = 3
```
I was able to lock workspaces with `iamPolicyVersion` set to 1, but when unlocking I got an error that the policy version is 3 and I can't use a lower version. The bucket IAM policy version in google said it was 1 though, so I'm not sure what is going on 😞 

It seems like if we always use 3 we won't run into issues because it will use version 1 when necessary: https://cloud.google.com/iam/docs/allow-policies#valid

```If the allow policy does not contain any conditions, then IAM always returns a version 1 allow policy, regardless of the version number in the request.```

samanehsan replied:
A bucket with fast pass grants will have conditional policies (which use version 3), so maybe that's why I saw the error

samanehsan commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -716,6 +715,142 @@ class WorkspaceServiceSpec
+    Await.result(services.workspaceService.lockWorkspace(testData.workspaceNoSubmissions.toWorkspaceName), Duration.Inf)
+    verify(services.gcsDAO).updateBucketIamAllReaders(
+      GcsBucketName(testData.workspaceNoSubmissions.bucketName),
+      Set(WorkbenchEmail(billingOwnerEmail),
```
No matter what values I put in this set, the test passes 🤔

davidangb replied:
if you include the `iamPolicyVersion` argument in this assertion, it works as expected:
```scala
    verify(services.gcsDAO).updateBucketIamAllReaders(
      GcsBucketName(testData.workspaceNoSubmissions.bucketName),
      Set(WorkbenchEmail(billingOwnerEmail),
          WorkbenchEmail(workspaceOwnersEmail),
          WorkbenchEmail(workspaceWritersEmail),
          WorkbenchEmail(workspaceReadersEmail)
      ),
      Some(testData.workspaceNoSubmissions.googleProjectId),
      iamPolicyVersion = 3
    )
```

TIL ... apparently Mockito does not work well with default arguments in Scala methods

samanehsan replied:
Awesome! Thanks for taking a look at that!!

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1603,13 +1603,55 @@ class WorkspaceService(
+        }
+      }
+      .flatten
+      .toMap
```
looks like this is copied from `createNewWorkspaceContext` ... how about extracting this into a reusable method somewhere for both locations to call? It feels like this would be too easy to get out of sync between the two places.

davidangb commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -716,6 +715,142 @@ class WorkspaceServiceSpec
+          WorkbenchEmail(workspaceWritersEmail),
+          WorkbenchEmail(workspaceReadersEmail)
+        ),
+        Some(testData.workspaceNoSubmissions.googleProjectId)
```
you'll have to update this `verify` as well

davidangb commented on core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:
```diff
@@ -716,6 +715,142 @@ class WorkspaceServiceSpec
+      GcsBucketName(testData.workspaceNoSubmissions.bucketName),
+      expectedPolicyEmails,
+      Some(testData.workspaceNoSubmissions.googleProjectId)
+    )
```
`updateBucketIam` also has a default argument, need to check these `verify` calls too!

samanehsan committed:
Fix mocking unit tests

samanehsan committed:
Scalafmt

samanehsan committed:
Refactor getting policy emails based on auth domain

samanehsan commented:
I finished making those changes @davidangb :octocat:

samanehsan committed:
Fix checking if auth domain is empty

davidangb reviewed (approved):
:+1: once tests pass, thank you!

samanehsan committed:
Fix checking if auth domain is empty when creating workspace

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-671-update-lock-logic

samanehsan committed:
Scalafmt
